Add introductory kernel notebook and change style file path in notebooks

Thomas-Christie opened the pull request:
## Type of changes

- [ ] Bug fix
- [ ] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [X] Other - New notebook.

## Checklist

- [X] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [X] I've added tests for new code.
- [X] I've added docstrings for the new code.

## Description

Main change is to add a notebook introducing the concept of a kernel for those new to Gaussian processes. This focuses on mathematical intuition and introduces useful concepts such as covariance matrices and positive-definiteness.

Also noted that the README file for writing documentation hadn't been updated since switching to using MkDocs, so updated it to reflect these changes.

Also found that the relative path to the style file for the notebooks caused issues when running the `poetry run mkdocs serve` command. Now instead of using a relative path to the style file, we point to the URL of the style file directly. I have copied the style file into the `_static` directory, and a future PR will point to this URL instead.

Finally made a few other minor edits:
- Updated docstring for the periodic kernel as it was incomplete.
- Made some minor fixes to a few other docstrings I found flaws in.
- Bumped up KaTeX version from 0.9.0 to 0.16.8 in `mkdocs.yml` file.

Issue Number: #269

thomaspinder commented on docs/examples/README.md:
```diff
@@ -53,16 +55,13 @@ class Prior(AbstractPrior):
-    .. math::
-
-        p(f(\\cdot)) = \mathcal{GP}(m(\\cdot), k(\\cdot, \\cdot)).
+    A Gaussian process prior parameterised by a mean function $`m(\cdot)`$ and a kernel
+    function $`k(\cdot, \cdot)`$ is given by
+    $`p(f(\cdot)) = \mathcal{GP}(m(\cdot), k(\cdot, \cdot))`$.
```
Why the backticks here?

Thomas-Christie replied:
Since this is an example of a docstring, I have included backticks as they're necessary for maths in docstrings to be formatted correctly.

thomaspinder commented on docs/examples/README.md:
```diff
@@ -53,16 +55,13 @@ class Prior(AbstractPrior):
-    To invoke a ``Prior`` distribution, only a kernel function is required. By default,
-    the mean function will be set to zero. In general, this assumption will be reasonable
-    assuming the data being modelled has been centred.
+    To invoke a `Prior` distribution, only a kernel function is required. By
+    default, the mean function will be set to zero. In general, this assumption
```
Not true anymore actually. The user must specify a mean function too.

Thomas-Christie replied:
Ok, will change (as well as in the original docstring)

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+# %% [markdown]
+# Using Gaussian Processes (GPs) to model functions can offer several advantages over alternative methods, such as deep neural networks. One key advantage is their rich quantification of uncertainty; not only do they provide *point estimates* for the values taken by a function throughout its domain, but they provide a full predictive posterior *distribution* over the range of values the function  ...
+#
+# However, another advantage of GPs is the ability for one to place *priors* on the functions being modelled. For instance, one may know that the underlying function being modelled observes certain characteristics, such as being *periodic* or having a certain level of *smoothness*. The *kernel*, or *covariance function*, is the primary means through which one is able to encode such prior knowledg ...
```
Very nice :)

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+# ## Introducing a Common Family of Kernels - The Matérn Family
+
+# %% [markdown]
+# Intuitively, the kernel defines the notion of *similarity* between the value taken at two points, $\mathbf{x}$ and $\mathbf{x}'$, by a function $f$, and will be denoted as $k(\mathbf{x}, \mathbf{x}')$:
```
Not true for a stationary kernel - the similarity is calculated on the difference between two points, not the value taken at two points.

Thomas-Christie replied:
What I mean with this sentence is that for two inputs, $\mathbf{x}$ and $\mathbf{x}'$, the *output* of the kernel function is a measure of the similarity between the function value at those inputs, which I believe is a general property of kernels, so holds for stationary kernels? Are you happy with the following rewording to avoid ambiguity:

"Intuitively, for a function $f$, the kernel defines the notion of *similarity* between the value of the function at two points, $f(\mathbf{x})$ and $f(\mathbf{x}')$, and will be denoted as $k(\mathbf{x}, \mathbf{x}')$"

thomaspinder replied:
Yes, I'm good with that.

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+#
+# $$k(\mathbf{x}, \mathbf{x}') = \text{Cov}[f(\mathbf{x}), f(\mathbf{x}')]$$
+#
+#  One would expect that, given a previously unobserved test point $\mathbf{x}^*$, training points which are *closest* to this unobserved point will be most similar to it. As such, the kernel is used to define this notion of similarity within the GP framework. It tends to be up to the user to select a kernel which is appropriate for the function being modelled.
```
the training points which..

thomaspinder replied:
It is up to the user to select...

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+#
+#  One would expect that, given a previously unobserved test point $\mathbf{x}^*$, training points which are *closest* to this unobserved point will be most similar to it. As such, the kernel is used to define this notion of similarity within the GP framework. It tends to be up to the user to select a kernel which is appropriate for the function being modelled.
+#
+# One of the most widely used families of kernels is the Matérn family. These kernels take on the following form:
```
Cite Matérn here - the reference is in `gpjax/citation` if you need it.

Thomas-Christie replied:
Do you want me to cite with a hyperlink? It was my understanding that you can't cite in `.ipynb` files? Do let me know if there is a way though :)

thomaspinder replied:
Yes - a hyperlink please. Unfortunately, it's not possible to use citations in mkdocs/jupyter. If you want to be fancy, you can use `gpx.cite`, but I'll leave that one up to you :)

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+#
+# One of the most widely used families of kernels is the Matérn family. These kernels take on the following form:
+#
+# $$k_{\nu}(\mathbf{x}, \mathbf{x'}) = \sigma^2 \frac{2^{1 - \nu}}{\Gamma(\nu)}\left(\sqrt{2\nu} \frac{|\mathbf{x} - \mathbf{x'}|}{\kappa}\right)^{\nu} K_{\nu} \left(\sqrt{2\nu} \frac{|\mathbf{x} - \mathbf{x'}|}{\kappa}\right)$$
```
Is it worth giving the special cases where nu is half-integer valued?

Thomas-Christie replied:
Will give an example for Matérn32

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+# ## Inferring Kernel Hyperparameters
+
+# %% [markdown]
+# Most kernels have several *hyperparameters*, which we denote $\mathbf{\theta}$, which encode different assumptions about the underlying function being modelled. For the Matérn family descibred above, $\mathbf{\theta} = \{\nu, \kappa, \sigma\}$. A fully Bayesian approach to dealing with hyperparameters would be to place a prior over them, and marginalise over the posterior derived from the data  ...
```
Maybe note in gpjax that `\nu` is fixed and not something that can be inferred through optimisation.

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+#
+# $$ \begin{aligned}
+# \log(p(\mathbf{y} | \mathbf{X}, \boldsymbol{\theta})) &= \log\left(\int p(\mathbf{y} | \mathbf{f}, \mathbf{X}, \boldsymbol{\theta}) p(\mathbf{f} | \mathbf{X}, \boldsymbol{\theta}) d\mathbf{f}\right) \nonumber \\
+# &= - \frac{1}{2} \mathbf{y} ^ \top \left(K(\mathbf{X}, \mathbf{X}) + \sigma^2 \mathbf{I} \right)^{-1} \mathbf{y} - \frac{1}{2} \log |K(\mathbf{X}, \mathbf{X}) + \sigma^2 \mathbf{I}| - \frac{n}{2} \log 2 \pi
```
Maybe add that for more details, see the intro to GPs notebook. It's good to cross-reference docs.

Thomas-Christie replied:
Good idea, have linked.

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+
+# %%
+# Forrester function
+def forrester(x):
```
nit: types please

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+    key=key,
+)
+
+
```
A short markdown cell here stating that you'll now query the predictive posterior.

thomaspinder replied:
No need for maths, just break up a code block with some narration.

thomaspinder commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,328 @@
+# %% [markdown]
+# ## What are the Necessary Conditions for a Valid Kernel?
+
+# %% [markdown]
```
It might be nice to expand our this section a little to lower the bar-to-entry e.g., define a Gram matrix and PSD, then move to the start of the notebook.

thomaspinder replied:
Sure. This is your work and I'm happy for you to own the style. My comment was just a suggestion with the intention of making the notebook as accessible to new users as possible.

thomaspinder reviewed (changes requested):
Really nice work @Thomas-Christie! I've left comments.

nit: can you make all markdown lines 88 characters long please? It makes reviewing much easier. Most code editors will have a hard wrap option that should automate this for you.

Thomas-Christie commented:
Thanks a lot for the detailed feedback @thomaspinder . I have addressed your comments in a new commit.

Thomas-Christie commented:
Minor commit added - had forgotten to add the reference for the Matérn kernel.

thomaspinder commented:
Approved. I can see @henrymoss has added himself as a reviewer. Once his review cycle is complete, I am happy for this to be merged.

henrymoss commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,369 @@
+#  framework. It is up to the user to select a kernel which is appropriate for the
+#  function being modelled.
+#
+# One of the most widely used families of kernels is the Matérn family ([Matérn, 1960](https://core.ac.uk/download/pdf/11698705.pdf)). These kernels take on the following form:
```
I think maybe add a quick note that its not important to understand the maths here. Just so that lots of differet types of people can enjoy the notebook. Maybe say "if you are not iterested in the maths, skip to X to see how we can use different kernels to encode different assumptions"

henrymoss commented on docs/examples/intro_to_kernels.py:
```diff
@@ -0,0 +1,369 @@
+ax.plot(x, y.T, alpha=0.7)
+ax.set_title("Samples from the Linear Kernel")
+plt.show()
+
```
It would be great if you added a couple more examples. Perhaps show what happens when you add or multiple kernels together?

henrymoss replied:
I think it would be super cool to end with a recreation of the "Mauna Loa atmospheric CO2" example of Section 5 of https://proceedings.mlr.press/v28/duvenaud13.pdf

This is very much "on brand" with the environmental focus of GpJax and demonstrates loads of cool stuff, i.e. its a periodic/sum of multiple base kernels? What do you think @thomaspinder

henrymoss replied:
![image](https://github.com/JaxGaussianProcesses/GPJax/assets/32096840/108b7cb1-ce42-4de7-8850-1c8849d74056)

henrymoss replied:
The airline dataset is perhaps even more interesting from a modelling perspective 
![image](https://github.com/JaxGaussianProcesses/GPJax/assets/32096840/c08c4379-ee32-4a0b-8107-1ebecaa99521)

Thomas-Christie replied:
> It would be great if you added a couple more examples. Perhaps show what happens when you add or multiple kernels together?

I have now added a section on composing kernels through addition/multiplication of existing kernels.

Thomas-Christie replied:
> I think it would be super cool to end with a recreation of the "Mauna Loa atmospheric CO2" example of Section 5 of https://proceedings.mlr.press/v28/duvenaud13.pdf
> 
> This is very much "on brand" with the environmental focus of GpJax and demonstrates loads of cool stuff, i.e. its a periodic/sum of multiple base kernels? What do you think @thomaspinder

Nice idea, I now have a section where we fit the CO2 dataset with a custom kernel.

henrymoss reviewed (approved):
Great work Tom. Really nice noteobook. I've left a couple of comments. I think we might as well show off some Sum / Products or kernels here

henrymoss commented:
Also, now that we have an introductory kernel notebook, shall we rename the existing kernels.py to building_new_kernels.py or something similar?

Thomas-Christie commented:
> Also, now that we have an introductory kernel notebook, shall we rename the existing kernels.py to building_new_kernels.py or something similar?

I have renamed the file to `constructing_new_kernels.py` and all references to the file.

Thomas-Christie commented:
@henrymoss @thomaspinder thanks for the feedback. I have addressed Henry's feedback, adding a section using the Mauna Loa CO2 dataset and a section on composing new kernels through addition/multiplication of existing kernels. I have also moved the section on Gram matrices etc. closer to the start.

Have also renamed the original 'kernels.py' notebook to 'constructing_new_kernels.py' for clarity (and have renamed references to this file).

Are you happy for me to merge this now?
